### PR TITLE
refactor: versioned filter on component ignore

### LINKED ISSUE #96: versioned filter on component ignore

### PR DESCRIPTION
**Pull request by dayuy**
Fix #96 

Replace filter.operation with filter.enabled.

image

### CODE REVIEW

File: api/v1alpha1/common.go [hunk 1]
```diff
@@ -297,41 +297,34 @@ func Match(fc map[string]FilterCond, filter Filter, funcs ...FilterFunc) ([]int,
...
 		for _, f := range funcs {
 			if f(filterCond, v.Version) {
-				if filterCond.Operation == FilterOpKeep && (filterCond.Deprecated || !v.Deprecated) {
-					keep = true
-				}
-				if filterCond.Operation == FilterOpIgnore {
-					keep = false
-				}
+				keep = false
+				keep = false
```
Comment (0xff-dev) [inline api/v1alpha1/common.go:321]: delete one？

**Comment by Abirdcfly** [comment]
this logic picture (ascii version) can be added into code.

File: api/v1alpha1/repository_types.go [hunk 2]
```diff
@@ -54,10 +47,9 @@ type FilterCond struct {
 	// Name of the component
 	Name string `json:"name,omitempty"`
 
-	// default is keep
-	// +kubebuilder:validation:Enum=keep;ignore
-	// +kubebuilder:default:=keep
-	Operation FilterOp `json:"operation,omitempty"`
```
Comment (0xff-dev) [inline api/v1alpha1/repository_types.go:60]: run `make generate`, `make manifests`

File: api/v1alpha1/common.go [hunk 3]
```diff
@@ -297,41 +297,34 @@ func Match(fc map[string]FilterCond, filter Filter, funcs ...FilterFunc) ([]int,
...
 				break
 			}
-			if filterCond.Operation == FilterOpIgnore && (filterCond.Deprecated || !v.Deprecated) {
-				keep = true
-			}
 		}
-
+		if !filterCond.Deprecated && v.Deprecated {
```
Comment (0xff-dev) [inline api/v1alpha1/common.go:325]: Perhaps we could define FilterFunc as follows:
```go
type FilterFunc func(FilterCond, *hrepo.ChartVersion) bool   
```
This judgment can also be turned into a function.

**Review by 0xff-dev — APPROVED** [review]

File: api/v1alpha1/repository_types.go [hunk 4]
```diff
@@ -54,10 +47,9 @@ type FilterCond struct {
...
+	// Enable filtering on the component,
+	// if false, no action will be taken on the component, in other worlds, all versions of the component will be retained.
+	Enabled bool `json:"enabled,omitempty"`
 
 	// If True, the current version will be retained even if it is deprecated.
 	Deprecated bool `json:"deprecated,omitempty"`
```
Comment (Abirdcfly) [inline api/v1alpha1/repository_types.go:55]: To make the usage clearer, how about replacing `deprecated` with `keepDeprecated`?

**Review by Abirdcfly — APPROVED** [review]
Others are fine.Thanks @dayuy

**Comment by dayuy** [comment]
Hold on, it seems not suitable for the PM‘s design.
image

**Comment by bjwswang** [comment]
@0xff-dev @dayuy If filter changed, say we removed a `filterCond`. Do we need to restart the `watcher`?

**Comment by bjwswang** [comment]
Great! @0xff-dev @dayuy  This makes the version fileter much clear!

File: api/v1alpha1/repository_types.go (hunk 4 above)
Comment (bjwswang) [reply api/v1alpha1/repository_types.go:55]: @dayuy pls change `deprecated` to `keepDeprecated`

**Comment by 0xff-dev** [comment]
> @0xff-dev @dayuy If filter changed, say we removed a `filterCond`. Do we need to restart the `watcher` to make the new `filterConds` take effects?

When the filter is updated, the watcher is automatically restarted

**Pull request merged**